Update to column handling when moving between rows

big-due opened the pull request:
I made some changes to help keep the cursor moving as expected when going between rows.

The issue that still exists in this fix but in a more specific place:  the bottom row 'keys' are much wider but are still treated the same as the single letter keys in my change.  Because of this, when moving between the bottom row and the lowest row with letters, even though 'b' and 'n' are directly above 'space', moving down from 'n' shifts over to 'enter' rather than directly down.  I considered targeting this further but it felt like anything to handle this specifically might make it a nightmare if/when a digits layout is implemented.

fixes #4

josegonzalez commented:
I think we wouldn't need to worry about shift for a digits keyboard, since it would work _only_ in digit mode. That said, I'm struggling to think right now of a digits use-case, so maybe I won't implement that ever?

josegonzalez commented:
If you want to work on the last row issue in this PR, great! If not, I can test/merge as is :)

big-due commented:
I had similar thoughts regarding the digits layout.  The standard layout having the number bar visible by default covers that IMO, but I could be convinced either way on that one.

I'll have another look at this tonight to see about the last lingering thing I noted and get back to you tonight or tomorrow if that works for you.

josegonzalez commented:
Sounds good!

big-due commented:
Alright, I think I have this done. While working on it I found a case where switching layouts could leave the cursor off-screen if it was on the rightmost key of a row and the new layout had fewer keys. get_current_layout and cursor_rescue help mitigate this, and I made a few targeted tweaks to use this right at the top of handle_keyboard_input and within it when the 'select' key is pressed.

The majority of the changes are the added helper functions and the logic in the handling of BTN_UP and BTN_DOWN.

Changes:
- Added functions to calculate proper cursor positioning when moving between rows
- Added special handling for bottom row navigation due to wider special keys
- Fixed cursor getting stuck off-screen when switching to layouts with fewer keys
- Refactored cursor boundary checking into a reusable rescue function

New functions:
- count_row_length: Counts non-empty characters in a keyboard row
- calculate_column_offset: Calculates position offset when moving between rows
- adjust_offset_exit_last_row: Adjusts offset when moving from bottom row
- adjust_offset_enter_last_row: Adjusts offset when moving to bottom row  
- get_current_layout: Returns current keyboard layout (refactored from existing code)
- cursor_rescue: Ensures cursor stays within valid positions

I've tested this with several modified keyboard layouts with different numbers of keys and the behavior has been good across all of them. There were a few places where I just had to use my best judgement regarding where I'd expect the cursor to move, mostly regarding getting into and out of the bottom row. The biggest example of this is that pressing up from 5, or 6, or 7 will all roll into 'space', and a similar behavior rolling from the bottom row keys to the top row.

josegonzalez commented:
Annoying that we can't prove the provenance of the files for PRs...
